### DISCUSSION TITLE
Amend an existing Type.Object

### DISCUSSION THREAD

**Discussion by wyozi** [sinclairzx81/typebox#586] [Q&A]
Hey, would it be possible to amend an existing type object in a stable way?

I mean something like 

```ts
const first = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
});
const second = Type.Amend(first, {
  z: Type.Number(),
})

// second is equal to 
const T = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
  z: Type.Number(),
});
```

Seems like a bit of a weird usecase, but it's for a library concept I'm prototyping.

**Comment by sinclairzx81** [comment]
@wyozi Hi, Does this work?

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/LAKAlgtgDg9gTgFwAQG8kBUCeUCmAaJAZQQEMEwBjA9AeQCMArHChagBThl0TBwGckAXyQAzThCQByAAJ8wAOwoAbEmDgB6BNhx0YAD0mhQOPbERIKMeX2QBBCDnkATJAF4kAHnRITCR04FaRmZWJDYfPT9nQI4uHB5+AD4ACj4KAAscCBIALgwCKE5ucn48tgBKN0SkUCQMbQA6AGEYaBg5P2SAbTTM7OpG+iYWZMK4hL5ygF1yoxBLa2QRNRs3etwGoZDklFqkPTysDYA5AFcIOnjk8rw9zEPGs4ur2ZBBV9AF1b5mKxd3ez+ZLLOA2Ai7EB1ABeDxO50ucGuoHecy0uAwa2IZEoHjROBgIiQPwWTkSQA)

```typescript
import { Type, Static, TObject, TProperties } from '@sinclair/typebox'

export const Amend = <T extends TObject, P extends TProperties>(schema: T, properties: P) => 
  Type.Composite([schema, Type.Object(properties)])

const first = Type.Object({
  x: Type.Number(),
  y: Type.Number()
})

const second = Amend(first, {
  z: Type.Number()
})

type T = Static<typeof second>
```

**Reply by wyozi** [reply]
@sinclairzx81 thanks! This works very well for additive partial schemas! :)

This is already enough for most cases, but just wondering would there be a way to do deletions/replacements as well?

Namely,

```ts
const third = Amend(second, {
  z: Type.Never()
}) // -> { x: number; y: number; }

const third = Amend(second, {
  z: Type.String
}) // -> { x: number; y: number; z: string }
```

**Reply by sinclairzx81** [reply]
@wyozi Heya. TypeBox is built to mirror TypeScript and supports most of the built in utility types. You can compose those additional types in the following way.

[TypeScript Link Here](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAbzgFQJ5gKYBoUHkBGAVhgMYw7IAKUEmswGAznAL5wBmNIcA5AAKNgAOxIAbAIbAoAehjoM+CAA8eAWABQGjEsiw4JCEMbwAgiAxCAJnAC8cADzI42mBcvNkBYmRyVnS1ysPaloMeiYAPgAKRhIACwwQcQAuFBwwGjoYBkZUygBKWwi4DTgUeQA6AGEIcAhBVyiAbViEpIpKr1IYKIzQ8MZ8gF18jQ0DI3h2KWNbcswKrrIohFK4JVS0BYA5AFcQfDCo-Kw11E3KvYOj0fUWW-HDWcZSQ2s7Mzco6ahjHFX1GUAF4XHb7Q5QY4ae5jdQTWYwOJSd7zDCLEDAHovCaWHBNHhAngjMok0llaTSOCWDCiDCuOAguBCcFhWHwqYQXawOJzT5WKKI5H-Mki0XkyniSzWRnGKDCADma0ZWzRAGUYHKhPKoXdbkA)

```typescript
import { Type, TObject, TProperties } from '@sinclair/typebox'

export const Amend = <T extends TObject, P extends TProperties>(schema: T, properties: P) => 
  Type.Composite([schema, Type.Object(properties)])

const first = Type.Object({
  x: Type.Number(),
  y: Type.Number()
})

const second = Amend(first, {
  z: Type.Number()
})

const third = Type.Omit(second, ['z'])           // delete z: number

const fourth = Amend(third, {                    // add z: string
  z: Type.String()
})

```
Have a read over https://www.typescriptlang.org/docs/handbook/utility-types.html for a bit more information on TypeScript utility types (specifically Pick and Omit)
